From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Brandon Williams <turbine.bmw@gmail.com>
Date: Sun, 19 Jul 2026 17:32:14 -0400
Subject: [PATCH] arm64: dts: qcom: enable Surface Pro 11 cameras

Describe the camera PMIC rails, clocks, resets, CCI buses, and CAMSS
endpoints for the VD55G0 IR, OV13858 rear, and IMX681 front camera
modules.

The board description is based on standard firmware enumeration and
hardware observations made on the author's Surface Pro 11.

Signed-off-by: Brandon Williams <turbine.bmw@gmail.com>
---
 .../boot/dts/qcom/x1-microsoft-denali.dtsi    | 301 ++++++++++++++++++
 1 file changed, 301 insertions(+)

diff --git a/arch/arm64/boot/dts/qcom/x1-microsoft-denali.dtsi b/arch/arm64/boot/dts/qcom/x1-microsoft-denali.dtsi
index 5a32f393a704..83376d84fb2d 100644
--- a/arch/arm64/boot/dts/qcom/x1-microsoft-denali.dtsi
+++ b/arch/arm64/boot/dts/qcom/x1-microsoft-denali.dtsi
@@ -8,6 +8,7 @@
 #include <dt-bindings/input/gpio-keys.h>
 #include <dt-bindings/input/input.h>
 #include <dt-bindings/interrupt-controller/irq.h>
+#include <dt-bindings/media/video-interfaces.h>
 #include <dt-bindings/regulator/qcom,rpmh-regulator.h>
 
 #include "hamoa-pmics.dtsi"
@@ -495,6 +496,14 @@ vreg_l6b_1p8: ldo6 {
 			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
 		};
 
+		/* Front camera (IMX681) analog rail. */
+		vreg_l7b_2p8: ldo7 {
+			regulator-name = "vreg_l7b_2p8";
+			regulator-min-microvolt = <2800000>;
+			regulator-max-microvolt = <2800000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
 		vreg_l8b_3p0: ldo8 {
 			regulator-name = "vreg_l8b_3p0";
 			regulator-min-microvolt = <3072000>;
@@ -743,12 +752,256 @@ vreg_l3j_0p8: ldo3 {
 			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
 		};
 	};
+
+	/*
+	 * PM8010 "m": camera PMIC supplying the rear, front and IR modules.
+	 */
+	regulators-8 {
+		compatible = "qcom,pm8010-rpmh-regulators";
+		qcom,pmic-id = "m";
+
+		/*
+		 * TODO: supply parents are unverified. Chosen by voltage
+		 * headroom from the rails present on this board, following
+		 * the PM8010 wiring seen on other X1E80100 designs
+		 * (1.8 V LDOs from vreg_s4c_1p8, 2.8 V LDOs from vreg_bob1).
+		 */
+		vdd-l1-l2-supply = <&vreg_s5j_1p2>;
+		vdd-l3-l4-supply = <&vreg_s4c_1p8>;
+		vdd-l5-supply = <&vreg_bob1>;
+		vdd-l6-supply = <&vreg_s4c_1p8>;
+		vdd-l7-supply = <&vreg_bob1>;
+
+		vreg_l1m_1p2: ldo1 {
+			regulator-name = "vreg_l1m_1p2";
+			regulator-min-microvolt = <1200000>;
+			regulator-max-microvolt = <1200000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		/* Front camera (IMX681) IO rail. */
+		vreg_l3m_1p8: ldo3 {
+			regulator-name = "vreg_l3m_1p8";
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l2m_1p15: ldo2 {
+			/* 1.152 V is the closest PM8010 step to the 1.15 V rail. */
+			regulator-name = "vreg_l2m_1p15";
+			regulator-min-microvolt = <1152000>;
+			regulator-max-microvolt = <1152000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l4m_1p8: ldo4 {
+			regulator-name = "vreg_l4m_1p8";
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l5m_2p8: ldo5 {
+			regulator-name = "vreg_l5m_2p8";
+			regulator-min-microvolt = <2800000>;
+			regulator-max-microvolt = <2800000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l6m_1p8: ldo6 {
+			regulator-name = "vreg_l6m_1p8";
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+
+		vreg_l7m_2p8: ldo7 {
+			regulator-name = "vreg_l7m_2p8";
+			regulator-min-microvolt = <2800000>;
+			regulator-max-microvolt = <2800000>;
+			regulator-initial-mode = <RPMH_REGULATOR_MODE_HPM>;
+		};
+	};
+};
+
+&camcc {
+	status = "okay";
+};
+
+&camss {
+	vdd-csiphy-0p8-supply = <&vreg_l2c_0p8>;
+	vdd-csiphy-1p2-supply = <&vreg_l1c_1p2>;
+
+	status = "okay";
+
+	ports {
+		/*
+		 * port@0 => CSIPHY0: IR camera (ST VD55G0, 1-lane D-PHY)
+		 * port@1 => CSIPHY1: rear camera (OV13858, 4-lane D-PHY)
+		 * port@2 => CSIPHY2: front camera (Sony IMX681, 1-trio C-PHY),
+		 *           MCLK4 (gpio100 "cam_aon") @ 19.2 MHz, reset GPIO
+		 *           237, rails LDO3_M 1.8 V + LDO7_B 2.8 V.
+		 */
+		port@0 {
+			csiphy0_ep: endpoint {
+				clock-lanes = <7>;
+				data-lanes = <0>;
+				remote-endpoint = <&vd55g0_ep>;
+			};
+		};
+
+		port@1 {
+			csiphy1_ep: endpoint {
+				clock-lanes = <7>;
+				data-lanes = <0 1 2 3>;
+				remote-endpoint = <&ov13858_ep>;
+			};
+		};
+
+		port@2 {
+			csiphy2_ep: endpoint {
+				bus-type = <MEDIA_BUS_TYPE_CSI2_CPHY>;
+				data-lanes = <0>;
+				remote-endpoint = <&imx681_ep>;
+			};
+		};
+	};
+};
+
+&cci0 {
+	status = "okay";
+};
+
+&cci0_i2c0 {
+	/*
+	 * IR camera: ST VD55G0 (ACPI SMO55F0, MSHW0492).
+	 *
+	 * Hardware observations and firmware enumeration identify MCLK0 at
+	 * 19.2 MHz (gpio96), reset TLMM 109 active-low, rails LDO4_M 1.8 V
+	 * (vddio) -> LDO2_M 1.15 V (vcore) -> LDO7_M 2.8 V (vana). The
+	 * gpio101/102 CCI pair floated during the rear-camera bus hunt while
+	 * the rear rails were up, consistent with it being the (unpowered)
+	 * IR camera pair. The VD55G0 answers at 0x60 (verified on hardware;
+	 * unlike the VD55G1 whose address is fixed at 0x10). 1-lane
+	 * D-PHY into CSIPHY0 at 840 Mbps (link-frequency 420 MHz),
+	 * 644x604 RAW10 (Y10P) at approximately 30 fps.
+	 */
+	camera@60 {
+		compatible = "st,vd55g0";
+		reg = <0x60>;
+
+		reset-gpios = <&tlmm 109 GPIO_ACTIVE_LOW>;
+		pinctrl-0 = <&cam_ir_default>;
+		pinctrl-names = "default";
+
+		clocks = <&camcc CAM_CC_MCLK0_CLK>;
+		assigned-clocks = <&camcc CAM_CC_MCLK0_CLK>;
+		assigned-clock-rates = <19200000>;
+
+		VCORE-supply = <&vreg_l2m_1p15>;
+		VDDIO-supply = <&vreg_l4m_1p8>;
+		VANA-supply = <&vreg_l7m_2p8>;
+
+		orientation = <0>;	/* Front-facing */
+
+		port {
+			vd55g0_ep: endpoint {
+				data-lanes = <1>;
+				link-frequencies = /bits/ 64 <420000000>;
+				remote-endpoint = <&csiphy0_ep>;
+			};
+		};
+	};
+};
+
+&cci0_i2c1 {
+	/*
+	 * Rear camera: OmniVision OV13858 (ACPI OVTID858, MSHW0491).
+	 *
+	 * Bus + address verified on hardware 2026-07-11: with the sensor
+	 * rails, MCLK1 and reset held active, the sensor ACKs at 0x10 on
+	 * the gpio103/104 CCI pair, while the gpio101/102 pair floats
+	 * (presumed IR camera pair, unpowered). 0x10 is the OV13858
+	 * default SCCB address (SID low).
+	 */
+	camera@10 {
+		compatible = "ovti,ov13858";
+		reg = <0x10>;
+
+		reset-gpios = <&tlmm 110 GPIO_ACTIVE_LOW>;
+		pinctrl-0 = <&cam_rear_default>;
+		pinctrl-names = "default";
+
+		clocks = <&camcc CAM_CC_MCLK1_CLK>;
+		assigned-clocks = <&camcc CAM_CC_MCLK1_CLK>;
+		assigned-clock-rates = <19200000>;
+
+		avdd-supply = <&vreg_l5m_2p8>;
+		dvdd-supply = <&vreg_l1m_1p2>;
+		dovdd-supply = <&vreg_l6m_1p8>;
+
+		orientation = <1>;	/* Rear-facing */
+
+		port {
+			ov13858_ep: endpoint {
+				data-lanes = <1 2 3 4>;
+				link-frequencies = /bits/ 64 <540000000 270000000>;
+				remote-endpoint = <&csiphy1_ep>;
+			};
+		};
+	};
 };
 
 &gpu {
 	status = "okay";
 };
 
+&cci1 {
+	status = "okay";
+};
+
+&cci1_i2c1 {
+	/*
+	 * Front camera: Sony IMX681 (ACPI SONY0681, MSHW0490). The bus is
+	 * cci1 master 1 on gpio235/236 ("aon_cci"
+	 * pads on the AON IO island, same arrangement as the Zenbook A14
+	 * front camera on &cci1_i2c1). Hardware observations identify MCLK4
+	 * at 19.2 MHz (gpio100, "cam_aon"), reset TLMM 237 active-low,
+	 * LDO3_M 1.8 V (IO) + LDO7_B 2.8 V (analog). Address 0x1a is the
+	 * Sony default, and was verified on the hardware.
+	 *
+	 */
+	camera@1a {
+		compatible = "sony,imx681";
+		reg = <0x1a>;
+
+		reset-gpios = <&tlmm 237 GPIO_ACTIVE_LOW>;
+		pinctrl-0 = <&cam_front_default>;
+		pinctrl-names = "default";
+
+		clocks = <&camcc CAM_CC_MCLK4_CLK>;
+		assigned-clocks = <&camcc CAM_CC_MCLK4_CLK>;
+		assigned-clock-rates = <19200000>;
+
+		avdd-supply = <&vreg_l7b_2p8>;
+		dovdd-supply = <&vreg_l3m_1p8>;
+
+		orientation = <0>;	/* Front-facing */
+		rotation = <0>;
+
+		port {
+			imx681_ep: endpoint {
+				bus-type = <MEDIA_BUS_TYPE_CSI2_CPHY>;
+				data-lanes = <0>;
+				/* Half of the observed 2406-Msym/s C-PHY symbol rate. */
+				link-frequencies = /bits/ 64 <1203000000>;
+				remote-endpoint = <&csiphy2_ep>;
+			};
+		};
+	};
+};
+
 &gpu_zap_shader {
 	firmware-name = "qcom/x1e80100/microsoft/qcdxkmsuc8380.mbn";
 };
@@ -1234,6 +1487,54 @@ ssam_state: ssam-state-state {
 		bias-disable;
 	};
 
+	cam_front_default: cam-front-default-state {
+		mclk-pins {
+			pins = "gpio100";
+			function = "cam_aon";
+			drive-strength = <16>;
+			bias-disable;
+		};
+
+		reset-n-pins {
+			pins = "gpio237";
+			function = "gpio";
+			drive-strength = <2>;
+			bias-disable;
+		};
+	};
+
+	cam_ir_default: cam-ir-default-state {
+		mclk-pins {
+			pins = "gpio96";
+			function = "cam_mclk";
+			drive-strength = <16>;
+			bias-disable;
+		};
+
+		reset-n-pins {
+			pins = "gpio109";
+			function = "gpio";
+			drive-strength = <2>;
+			bias-disable;
+		};
+	};
+
+	cam_rear_default: cam-rear-default-state {
+		mclk-pins {
+			pins = "gpio97";
+			function = "cam_mclk";
+			drive-strength = <16>;
+			bias-disable;
+		};
+
+		reset-n-pins {
+			pins = "gpio110";
+			function = "gpio";
+			drive-strength = <2>;
+			bias-disable;
+		};
+	};
+
 	wcn_wlan_bt_en: wcn-wlan-bt-en-state {
 		pins = "gpio116", "gpio117";
 		function = "gpio";
